pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+         */
+        void close() {
+            if (mode_ == Mode_::UNINIT) {
+                return;
+            }
+            else if (mode_ == Mode_::READONLY) {
+                delete mmap_r_;
+                mmap_r_ = nullptr;
+            }
+            else if (mode_ == Mode_::WRITEONLY) {
+                delete mmap_w_;
+                mmap_w_ = nullptr;
+            }
+            mode_ = Mode_::UNINIT;
+        }
+
+        /* Expose get/put functions */
+        template<class T, class TT=T, details_::if_simple_serializable<TT> = true>
+        T fget_one() {
+            check_get_();
+            return mmap_r_->fget_one<T>();
+        }
+
+        template<class T, class TT=T, details_::if_simple_serializable<TT> = true>
+        T* fget_multiple(const uint64_t n_elements) {
+            check_get_();
+            return mmap_r_->fget_multiple<T>(n_elements);
+        }
+
+        template<class T, class TT=T, details_::if_simple_serializable<TT> = true>
+        void fput_one(const T& src) {
+            check_put_();
+            mmap_w_->fput_one<T>(src);
+        }
+
+        template<class T, class TT=T, details_::if_simple_serializable<TT> = true>
+        void fput_multiple(const T* src, const uint64_t n_elements) {
+            check_put_();
+            mmap_w_->fput_multiple<T>(src, n_elements);
+        }
+
+    private:
+        enum class Mode_ {
+            UNINIT = 0,             // Uninitialized
+            READONLY = 1,           // "r"
+            WRITEONLY = 2,          // "w"
+        };
+
+        details_::MmapStoreSave_* mmap_w_ = nullptr; // Used for write mode
+        details_::MmapStoreLoad_* mmap_r_ = nullptr; // Used for read mode
+        Mode_ mode_ = Mode_::UNINIT; // Indicator for current opened mode
+
+        /* Check if opened as read mode for calling get functions */
+        void check_get_() {
+            if (!mmap_r_) {
+                throw std::runtime_error("Not opened for read mode, cannot call get.");
+            }
+        }
+
+        /* Check if opened as write mode for calling put functions */
+        void check_put_() {
+            if (!mmap_w_) {
+                throw std::runtime_error("Not opened for write mode, cannot call put.");
+            }
+        }
+}; // end class MmapStore
+
+
+/* Extended Vector class that can be used as a std::vector or as a memory-view from part of mmap file
+ * For std::vector case, it own the memory for data storage.
+ * For mmap view case, it does not own any memory, but serve as a view for a piece of memory owned by MmapStore.
+ * By default, it is initialized as empty std::vector that can be resized or loaded as mmap view.
+ * Once loaded as mmap view, it cannot go back to std::vector case unless clear() is called.
+ */
+template<class T, class TT = T, details_::if_simple_serializable<TT> = true>
+class MmapableVector {
+    public:
+        MmapableVector() {} // Dummy constructor for empty case
```

Comment:
As this class will be used as a drop-in replacement of std::vector with limited operations. Do we want to consider to follow the rule of five pattern for this class?  https://en.cppreference.com/w/cpp/language/rule_of_three